# Stroke data Exploration

## Loading


In [34]:
import pandas as pd

In [35]:
#drive.mount('/content/drive')

In [36]:
df = pd.read_csv('data/stroke.csv')
pd.set_option("display.precision", 2)

## Column values

### General exploration

In [37]:
df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

In [38]:
print(
    f" Type = {type(df)}\n Len = {len(df)}\n Shape = {df.shape} "
)

 Type = <class 'pandas.core.frame.DataFrame'>
 Len = 5110
 Shape = (5110, 12) 


In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


In [40]:
df.describe()


id      age  hypertension  heart_disease  avg_glucose_level  \
count   5110.00  5110.00        5110.0        5110.00            5110.00   
mean   36517.83    43.23           0.1           0.05             106.15   
std    21161.72    22.61           0.3           0.23              45.28   
min       67.00     0.08           0.0           0.00              55.12   
25%    17741.25    25.00           0.0           0.00              77.25   
50%    36932.00    45.00           0.0           0.00              91.88   
75%    54682.00    61.00           0.0           0.00             114.09   
max    72940.00    82.00           1.0           1.00             271.74   

           bmi   stroke  
count  4909.00  5110.00  
mean     28.89     0.05  
std       7.85     0.22  
min      10.30     0.00  
25%      23.50     0.00  
50%      28.10     0.00  
75%      33.10     0.00  
max      97.60     1.00

In [41]:
df.describe(include=object)

gender ever_married work_type Residence_type smoking_status
count     5110         5110      5110           5110           5110
unique       3            2         5              2              4
top     Female          Yes   Private          Urban   never smoked
freq      2994         3353      2925           2596           1892

In [42]:
df['stroke'] = df['stroke'].astype('category')

### Column exploration

In [43]:
df[df["id"].duplicated()]

Empty DataFrame
Columns: [id, gender, age, hypertension, heart_disease, ever_married, work_type, Residence_type, avg_glucose_level, bmi, smoking_status, stroke]
Index: []

No hi ha cap duplicat en la columna de id, com no aporta informació l'eliminarem

In [44]:
df.drop("id", axis=1, inplace=True)

In [45]:
df["gender"].value_counts()

Female    2994
Male      2115
Other        1
Name: gender, dtype: int64

Observem que hi ha una fila en que el genere del pacient no es home ni dona. Al ser una sola entrada de 5.110, l'eliminem del dataset

In [46]:
df.drop(df.index[df['gender']=="Other"], inplace=True)

In [47]:
df["age"].describe()

count    5109.00
mean       43.23
std        22.61
min         0.08
25%        25.00
50%        45.00
75%        61.00
max        82.00
Name: age, dtype: float64

Es pot observar que els valors màxim i mínim son coherents amb el que caldria esperar del rang d'edats (trobem valors decimals ja que també hi ha dades que correponen a infants)


In [48]:
df["hypertension"].value_counts()

0    4611
1     498
Name: hypertension, dtype: int64

In [49]:
df["heart_disease"].value_counts()

0    4833
1     276
Name: heart_disease, dtype: int64

In [50]:
df["ever_married"].value_counts()

Yes    3353
No     1756
Name: ever_married, dtype: int64

In [51]:
df["Residence_type"].value_counts()

Urban    2596
Rural    2513
Name: Residence_type, dtype: int64

Substitue binary types with 0 and 1

In [52]:
df.replace({'Yes':1, 'No':0, 'Urban':1, 'Rural':0, 'Male':0, 'Female':1}, inplace=True)

df['heart_disease'] = df['heart_disease'].astype('category')
df["hypertension"] = df["hypertension"].astype('category')
df["gender"] = df["gender"].astype('category')
df["ever_married"] = df["ever_married"].astype('category')
df["Residence_type"] = df["Residence_type"].astype('category')


In [53]:
df["work_type"].value_counts()

Private          2924
Self-employed     819
children          687
Govt_job          657
Never_worked       22
Name: work_type, dtype: int64

In [54]:
df["avg_glucose_level"].describe()

count    5109.00
mean      106.14
std        45.29
min        55.12
25%        77.24
50%        91.88
75%       114.09
max       271.74
Name: avg_glucose_level, dtype: float64

In [55]:
df["bmi"].describe()

count    4908.00
mean       28.89
std         7.85
min        10.30
25%        23.50
50%        28.10
75%        33.10
max        97.60
Name: bmi, dtype: float64

In [56]:
df["smoking_status"].value_counts()

never smoked       1892
Unknown            1544
formerly smoked     884
smokes              789
Name: smoking_status, dtype: int64

In [57]:
df["stroke"].value_counts()

0    4860
1     249
Name: stroke, dtype: int64

### NaN values

In [58]:
df.isna().any()

gender               False
age                  False
hypertension         False
heart_disease        False
ever_married         False
work_type            False
Residence_type       False
avg_glucose_level    False
bmi                   True
smoking_status       False
stroke               False
dtype: bool

In [59]:
df[df.isna().any(axis='columns') ].shape

(201, 11)

In [60]:
df[df.isna().any(axis='columns') & (df["stroke"] == 1)].shape

(40, 11)

In [61]:
df[(df["stroke"] == 1)].shape

(249, 11)

Contem el nombre de files que contenen valors NaN, trobem 201 del total de 5110. I d'aquests 201, 40 han tingut un derrame.

#### Creació de data set eliminant els valors nuls

In [62]:
df_noNulls = df.dropna()
print(f'df shape = {df.shape}\ndf_noNulls shape = {df_noNulls.shape} ')

df shape = (5109, 11)
df_noNulls shape = (4908, 11) 


In [63]:
# df_noNulls_zero = df_noNulls.query("stroke == 0").sample(n= 209)
# df_noNulls_one = df_noNulls[df_noNulls["stroke"]==1].copy()
# df_noNulls_reduced = pd.concat([df_noNulls_zero, df_noNulls_one])

## Guardem les dades

In [64]:
df.to_pickle('./data/clean_stroke.pkl')
df_noNulls.to_pickle('./data/noNulls.pkl')